File: clients/ios/Lextures/Core/LMS/TutorStreamClient.swift
import Foundation

/// Reads tutor/study-buddy SSE streams from POST endpoints.
struct TutorStreamClient {
    private let session: URLSession

    init(session: URLSession = NetworkBootstrap.makeSession()) {
        self.session = session
    }

    func stream(
        path: String,
        method: String = "POST",
        body: (any Encodable)?,
        accessToken: String
    ) -> AsyncThrowingStream<TutorStreamEvent, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    var request = URLRequest(url: AppConfiguration.apiURL(path: path))
                    request.httpMethod = method
                    request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    request.setValue("ios", forHTTPHeaderField: "X-Platform")
                    request.setValue(LocalePreferences.acceptLanguageHeaderValue(), forHTTPHeaderField: "Accept-Language")
                    if let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String {
                        request.setValue(version, forHTTPHeaderField: "X-App-Version")
                    }
                    request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
                    if let body {
                        request.httpBody = try JSONEncoder().encode(TutorRequestBody(body))
                    }

                    let (bytes, response) = try await session.bytes(for: request)
                    guard let http = response as? HTTPURLResponse else {
                        throw APIError.invalidResponse
                    }
                    if !(200 ... 299).contains(http.statusCode) {
                        var data = Data()
                        for try await byte in bytes {
                            data.append(byte)
                        }
                        let message = TutorLogic.gracefulHttpMessage(
                            statusCode: http.statusCode,
                            body: parseAPIErrorMessage(from: data)
                        )
                        throw APIError.httpStatus(http.statusCode, message: message)
                    }

                    for try await line in bytes.lines {
                        try Task.checkCancellation()
                        if let event = TutorLogic.parseSSELine(line) {
                            continuation.yield(event)
                            switch event {
                            case .done, .error:
                                continuation.finish()
                                return
                            case .content:
                                continue
                            }
                        }
                    }
                    continuation.finish()
                } catch is CancellationError {
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}

private struct TutorRequestBody: Encodable {
    private let value: any Encodable

    init(_ value: any Encodable) { self.value = value }

    func encode(to encoder: Encoder) throws {
        try value.encode(to: encoder)
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/TutorStreamClient.kt
package com.lextures.android.core.lms

import com.lextures.android.BuildConfig
import com.lextures.android.core.config.AppConfiguration

import com.lextures.android.core.network.ApiError
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.isActive
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.atomic.AtomicReference

/** Reads tutor/study-buddy SSE streams from POST endpoints. */
class TutorStreamClient(
    private val http: OkHttpClient = OkHttpClient(),
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    private val activeCall = AtomicReference<okhttp3.Call?>(null)

    fun cancel() {
        activeCall.getAndSet(null)?.cancel()
    }

    fun stream(
        path: String,
        method: String = "POST",
        body: String?,
        accessToken: String,
    ): Flow<TutorStreamEvent> = flow {
        val builder = Request.Builder()
            .url(AppConfiguration.apiUrl(path))
            .header("Accept", "text/event-stream")
            .header("Authorization", "Bearer $accessToken")
            .header("X-Platform", "android")
            .header("X-App-Version", BuildConfig.VERSION_NAME)
            .header("Accept-Language", java.util.Locale.getDefault().toLanguageTag())

        if (body != null) {
            builder
                .method(method, body.toRequestBody("application/json".toMediaType()))
                .header("Content-Type", "application/json")
        } else {
            builder.method(method, null)
        }

        val call = http.newCall(builder.build())
        activeCall.set(call)
        try {
            val response = call.execute()
            response.use {
                if (it.code !in 200..299) {
                    val responseBody = it.body?.string().orEmpty()
                    throw ApiError.HttpStatus(
                        it.code,
                        TutorLogic.gracefulHttpMessage(it.code, responseBody),
                    )
                }
                val source = it.body?.source() ?: return@flow
                while (currentCoroutineContext().isActive && !source.exhausted()) {
                    val line = source.readUtf8Line() ?: break
                    TutorLogic.parseSseLine(line, json)?.let { event ->
                        emit(event)
                        if (event is TutorStreamEvent.Done || event is TutorStreamEvent.Error) return@flow
                    }
                }
            }
        } catch (e: Exception) {
            if (!call.isCanceled()) throw e
        } finally {
            activeCall.compareAndSet(call, null)
        }
    }.flowOn(Dispatchers.IO)
}